backend/oasst_backend/prompt_repository.py

```diff
@@ -488,3 +496,178 @@ def close_task(self, post_id: str, allow_personal_tasks: bool = False):
+        tree = self.fetch_message_tree(message)
+        max_message = max(tree, key=lambda m: m.depth)
+        return self.trace_conversation(tree, max_message)
+
+    def fetch_message_with_max_children(self, message: Post | UUID) -> tuple[Post, list[Post]]:
+        tree = self.fetch_message_tree(message)
+        max_message = max(tree, key=lambda m: m.children_count)
+        return max_message, [m for m in tree if m.parent_id == max_message.id]
+
+    def query_messages(
+        self,
+        user_id: Optional[UUID] = None,
+        username: Optional[str] = None,
+        api_client_id: Optional[UUID] = None,
+        desc: bool = True,
+        limit: Optional[int] = 10,
+        start_date: Optional[datetime.datetime] = None,
+        end_date: Optional[datetime.datetime] = None,
+        only_roots: bool = False,
+        deleted: Optional[bool] = None,
+    ) -> list[Post]:
+        if not self.api_client.trusted and not api_client_id:
+            # Let unprivileged api clients query their own messages without api_client_id being set
+            api_client_id = self.api_client.id
+
+        if not self.api_client.trusted and api_client_id != self.api_client.id:
+            # Unprivileged api client asks for foreign messages
+            raise OasstError("Forbidden", OasstErrorCode.API_CLIENT_NOT_AUTHORIZED, HTTP_403_FORBIDDEN)
+
+        messages = self.db.query(Post)
+        if user_id:
+            messages = messages.filter(Post.person_id == user_id)
+        if username:
+            messages = messages.join(Person)
+            messages = messages.filter(Person.username == username)
+        if api_client_id:
+            messages = messages.filter(Post.api_client_id == api_client_id)
+
+        if start_date:
+            messages = messages.filter(Post.created_date >= start_date)
+        if end_date:
+            messages = messages.filter(Post.created_date < end_date)
+
+        if only_roots:
+            messages = messages.filter(Post.parent_id.is_(None))
+
+        if deleted is not None:
+            messages = messages.filter(Post.deleted == deleted)
+
+        if desc:
+            messages = messages.order_by(Post.created_date.desc())
+        else:
+            messages = messages.order_by(Post.created_date.asc())
+
+        if limit is not None:
+            messages = messages.limit(limit)
+
+        # TODO: Pagination could be great at some point
+        return messages.all()
+
+    def mark_messages_deleted(self, messages: Post | UUID | list[Post | UUID], recursive: bool = True):
+        if isinstance(messages, (Post, UUID)):
+            messages = [messages]
+
+        ids = []
+        for message in messages:
+            if isinstance(message, UUID):
+                ids.append(message)
+            elif isinstance(message, Post):
+                ids.append(message.id)
+            else:
+                raise OasstError("Server error", OasstErrorCode.SERVER_ERROR, HTTPStatus.INTERNAL_SERVER_ERROR)
```

Comment:
same here for the error-site code, I know it is a bit inconvenient to add unique error codes...

Reply:
Same as above, I'd leave it as is, we have too little information about what happened here.

Reply:
It's more a general idea to have unique codes for all code-lines that throw an OasstError (independent of types). I have tried different things and made good experience with the unique call-site id approach .. otherwise you can only get this information from a stack-trace which normally isn't shown to the client.